File: Java/A-PREP/Code Signal Industry/In-Memory Database/level4.java
import java.util.*;

public class InMemoryDatabase {
    // Map to store records (key -> {field -> {value, expirationTimestamp}})
    private Map<String, Map<String, FieldValue>> database;
    // TreeMap to store backups (timestamp -> backup state)
    private TreeMap<Long, Map<String, Map<String, FieldValue>>> backups;

    // Constructor to initialize the database
    public InMemoryDatabase() {
        database = new HashMap<>();
        backups = new TreeMap<>();
    }

    // Inner class to store value and expiration timestamp
    private static class FieldValue {
        String value;
        long expirationTimestamp; // -1 means no expiration

        FieldValue(String value, long expirationTimestamp) {
            this.value = value;
            this.expirationTimestamp = expirationTimestamp;
        }

        FieldValue copy() {
            return new FieldValue(this.value, this.expirationTimestamp);
        }
    }

    // Method to insert or update a field-value pair with a timestamp
    public String setAt(String key, String field, String value, long timestamp) {
        database.putIfAbsent(key, new HashMap<>());
        database.get(key).put(field, new FieldValue(value, -1)); // No expiration
        return "";
    }

    // Method to insert or update a field-value pair with a timestamp and TTL
    public String setAtWithTTL(String key, String field, String value, long timestamp, long ttl) {
        database.putIfAbsent(key, new HashMap<>());
        database.get(key).put(field, new FieldValue(value, timestamp + ttl));
        return "";
    }

    // Method to delete a field at a specific timestamp
    public String deleteAt(String key, String field, long timestamp) {
        if (!database.containsKey(key)) {
            return "false";
        }
        Map<String, FieldValue> record = database.get(key);
        if (!record.containsKey(field)) {
            return "false";
        }
        record.remove(field);
        return "true";
    }

    // Method to get the value of a field at a specific timestamp
    public String getAt(String key, String field, long timestamp) {
        if (!database.containsKey(key)) {
            return "";
        }
        Map<String, FieldValue> record = database.get(key);
        if (!record.containsKey(field)) {
            return "";
        }
        FieldValue fieldValue = record.get(field);
        if (fieldValue.expirationTimestamp != -1 && timestamp >= fieldValue.expirationTimestamp) {
            return "";
        }
        return fieldValue.value;
    }

    // Method to scan all fields of a record at a specific timestamp
    public String scanAt(String key, long timestamp) {
        if (!database.containsKey(key)) {
            return "";
        }
        Map<String, FieldValue> record = database.get(key);
        // Filter out expired fields
        Map<String, String> validFields = new HashMap<>();
        for (Map.Entry<String, FieldValue> entry : record.entrySet()) {
            if (entry.getValue().expirationTimestamp == -1 || timestamp < entry.getValue().expirationTimestamp) {
                validFields.put(entry.getKey(), entry.getValue().value);
            }
        }
        // Sort the valid fields lexicographically
        TreeMap<String, String> sortedFields = new TreeMap<>(validFields);
        // Build the result string
        return buildResultString(sortedFields);
    }

    // Method to scan fields of a record that start with a prefix at a specific timestamp
    public String scanByPrefixAt(String key, String prefix, long timestamp) {
        if (!database.containsKey(key)) {
            return "";
        }
        Map<String, FieldValue> record = database.get(key);
        // Filter out expired fields and fields that don't start with the prefix
        Map<String, String> validFields = new HashMap<>();
        for (Map.Entry<String, FieldValue> entry : record.entrySet()) {
            if ((entry.getValue().expirationTimestamp == -1 || timestamp < entry.getValue().expirationTimestamp) &&
                entry.getKey().startsWith(prefix)) {
                validFields.put(entry.getKey(), entry.getValue().value);
            }
        }
        // Sort the valid fields lexicographically
        TreeMap<String, String> sortedFields = new TreeMap<>(validFields);
        // Build the result string
        return buildResultString(sortedFields);
    }

    // Method to back up the database state at a specific timestamp
    public String backup(long timestamp) {
        // Create a deep copy of the database
        Map<String, Map<String, FieldValue>> backup = new HashMap<>();
        for (Map.Entry<String, Map<String, FieldValue>> recordEntry : database.entrySet()) {
            Map<String, FieldValue> recordCopy = new HashMap<>();
            for (Map.Entry<String, FieldValue> fieldEntry : recordEntry.getValue().entrySet()) {
                recordCopy.put(fieldEntry.getKey(), fieldEntry.getValue().copy());
            }
            backup.put(recordEntry.getKey(), recordCopy);
        }
        // Save the backup
        backups.put(timestamp, backup);
        // Count non-empty, non-expired records
        int count = 0;
        for (Map.Entry<String, Map<String, FieldValue>> recordEntry : database.entrySet()) {
            boolean hasValidFields = false;
            for (Map.Entry<String, FieldValue> fieldEntry : recordEntry.getValue().entrySet()) {
                if (fieldEntry.getValue().expirationTimestamp == -1 || timestamp < fieldEntry.getValue().expirationTimestamp) {
                    hasValidFields = true;
                    break;
                }
            }
            if (hasValidFields) {
                count++;
            }
        }
        return String.valueOf(count);
    }

    // Method to restore the database state from a backup
    public String restore(long timestamp, long timestampToRestore) {
        // Find the latest backup before or at timestampToRestore
        Long backupTimestamp = backups.floorKey(timestampToRestore);
        if (backupTimestamp == null) {
            return ""; // No backup found (should not happen as per the problem statement)
        }
        // Get the backup
        Map<String, Map<String, FieldValue>> backup = backups.get(backupTimestamp);
        // Clear the current database
        database.clear();
        // Restore the backup and recalculate TTLs
        for (Map.Entry<String, Map<String, FieldValue>> recordEntry : backup.entrySet()) {
            Map<String, FieldValue> record = new HashMap<>();
            for (Map.Entry<String, FieldValue> fieldEntry : recordEntry.getValue().entrySet()) {
                FieldValue fieldValue = fieldEntry.getValue();
                long remainingTTL = fieldValue.expirationTimestamp == -1 ? -1 : fieldValue.expirationTimestamp - backupTimestamp;
                long newExpirationTimestamp = remainingTTL == -1 ? -1 : timestamp + remainingTTL;
                record.put(fieldEntry.getKey(), new FieldValue(fieldValue.value, newExpirationTimestamp));
            }
            database.put(recordEntry.getKey(), record);
        }
        return "";
    }

    // Helper method to build the result string
    private String buildResultString(Map<String, String> map) {
        StringBuilder result = new StringBuilder();
        for (Map.Entry<String, String> entry : map.entrySet()) {
            if (result.length() > 0) {
                result.append(", ");
            }
            result.append(entry.getKey()).append("(").append(entry.getValue()).append(")");
        }
        return result.toString();
    }

    // Main method to test the functionality
    public static void main(String[] args) {
        InMemoryDatabase db = new InMemoryDatabase();

        // Test case 1: SET_AT_WITH_TTL operation
        System.out.println(db.setAtWithTTL("A", "B", "C", 1, 10)); // ""
        System.out.println(db.backup(3)); // "1"

        // Test case 2: SET_AT operation
        System.out.println(db.setAt("A", "D", "E", 4)); // ""
        System.out.println(db.backup(5)); // "1"

        // Test case 3: DELETE_AT operation
        System.out.println(db.deleteAt("A", "B", 8)); // "true"
        System.out.println(db.backup(9)); // "1"

        // Test case 4: RESTORE operation
        System.out.println(db.restore(10, 7)); // ""
        System.out.println(db.backup(11)); // "1"

        // Test case 5: SCAN_AT operation
        System.out.println(db.scanAt("A", 15)); // "B(C), D(E)"
        System.out.println(db.scanAt("A", 16)); // "D(E)"
    }
}

File: Python/CodeSignal/Industry Framework/In-Memory Storage/level4.py
import copy

class InMemoryDatabase:
    def __init__(self):
        self.records = {}  # key -> {field: (value, expiry_time or None)}
        self.backups = []  # List of (timestamp, records_snapshot)
    
    def _cleanup_expired(self, timestamp: int):
        """Remove expired fields"""
        keys_to_remove = []
        for key in self.records:
            fields_to_remove = []
            for field, (value, expiry) in self.records[key].items():
                if expiry is not None and timestamp >= expiry:
                    fields_to_remove.append(field)
            
            for field in fields_to_remove:
                del self.records[key][field]
            
            if not self.records[key]:
                keys_to_remove.append(key)
        
        for key in keys_to_remove:
            del self.records[key]
    
    def _count_non_empty_records(self, timestamp: int) -> int:
        """Count non-empty, non-expired records"""
        count = 0
        for key in self.records:
            has_valid_fields = False
            for field, (value, expiry) in self.records[key].items():
                if expiry is None or timestamp < expiry:
                    has_valid_fields = True
                    break
            if has_valid_fields:
                count += 1
        return count
    
    # Level 1 operations (backward compatibility)
    def set(self, key: str, field: str, value: str) -> str:
        if key not in self.records:
            self.records[key] = {}
        
        self.records[key][field] = (value, None)  # No expiry
        return ""
    
    def get(self, key: str, field: str) -> str:
        if key not in self.records or field not in self.records[key]:
            return ""
        
        value, expiry = self.records[key][field]
        return value
    
    def delete(self, key: str, field: str) -> str:
        if key not in self.records or field not in self.records[key]:
            return "false"
        
        del self.records[key][field]
        
        if not self.records[key]:
            del self.records[key]
        
        return "true"
    
    def scan(self, key: str) -> str:
        if key not in self.records:
            return ""
        
        fields = sorted(self.records[key].keys())
        result_parts = []
        for field in fields:
            value, expiry = self.records[key][field]
            result_parts.append(f"{field}({value})")
        
        return ", ".join(result_parts)
    
    def scan_by_prefix(self, key: str, prefix: str) -> str:
        if key not in self.records:
            return ""
        
        matching_fields = []
        for field in self.records[key]:
            if field.startswith(prefix):
                matching_fields.append(field)
        
        matching_fields.sort()
        result_parts = []
        for field in matching_fields:
            value, expiry = self.records[key][field]
            result_parts.append(f"{field}({value})")
        
        return ", ".join(result_parts)
    
    # Level 3 operations with timestamp
    def set_at(self, key: str, field: str, value: str, timestamp: int) -> str:
        self._cleanup_expired(timestamp)
        
        if key not in self.records:
            self.records[key] = {}
        
        self.records[key][field] = (value, None)  # No expiry
        return ""
    
    def set_at_with_ttl(self, key: str, field: str, value: str, timestamp: int, ttl: int) -> str:
        self._cleanup_expired(timestamp)
        
        if key not in self.records:
            self.records[key] = {}
        
        expiry_time = timestamp + ttl
        self.records[key][field] = (value, expiry_time)
        return ""
    
    def get_at(self, key: str, field: str, timestamp: int) -> str:
        self._cleanup_expired(timestamp)
        
        if key not in self.records or field not in self.records[key]:
            return ""
        
        value, expiry = self.records[key][field]
        if expiry is not None and timestamp >= expiry:
            return ""
        
        return value
    
    def delete_at(self, key: str, field: str, timestamp: int) -> str:
        self._cleanup_expired(timestamp)
        
        if key not in self.records or field not in self.records[key]:
            return "false"
        
        value, expiry = self.records[key][field]
        if expiry is not None and timestamp >= expiry:
            return "false"
        
        del self.records[key][field]
        
        if not self.records[key]:
            del self.records[key]
        
        return "true"
    
    def scan_at(self, key: str, timestamp: int) -> str:
        self._cleanup_expired(timestamp)
        
        if key not in self.records:
            return ""
        
        valid_fields = []
        for field, (value, expiry) in self.records[key].items():
            if expiry is None or timestamp < expiry:
                valid_fields.append(field)
        
        valid_fields.sort()
        result_parts = []
        for field in valid_fields:
            value, expiry = self.records[key][field]
            result_parts.append(f"{field}({value})")
        
        return ", ".join(result_parts)
    
    def scan_by_prefix_at(self, key: str, prefix: str, timestamp: int) -> str:
        self._cleanup_expired(timestamp)
        
        if key not in self.records:
            return ""
        
        matching_fields = []
        for field, (value, expiry) in self.records[key].items():
            if field.startswith(prefix) and (expiry is None or timestamp < expiry):
                matching_fields.append(field)
        
        matching_fields.sort()
        result_parts = []
        for field in matching_fields:
            value, expiry = self.records[key][field]
            result_parts.append(f"{field}({value})")
        
        return ", ".join(result_parts)
    
    # Level 4 operations - backup and restore
    def backup(self, timestamp: int) -> str:
        self._cleanup_expired(timestamp)
        
        # Create a deep copy of current state with remaining TTLs
        backup_records = {}
        for key, fields in self.records.items():
            backup_records[key] = {}
            for field, (value, expiry) in fields.items():
                if expiry is None or timestamp < expiry:
                    # Calculate remaining TTL
                    if expiry is None:
                        remaining_ttl = None
                    else:
                        remaining_ttl = expiry - timestamp
                    backup_records[key][field] = (value, remaining_ttl)
        
        # Remove empty records
        backup_records = {k: v for k, v in backup_records.items() if v}
        
        # Store backup
        self.backups.append((timestamp, copy.deepcopy(backup_records)))
        
        return str(len(backup_records))
    
    def restore(self, timestamp: int, timestamp_to_restore: int) -> str:
        self._cleanup_expired(timestamp)
        
        # Find the latest backup before or at timestamp_to_restore
        backup_to_use = None
        for backup_timestamp, backup_data in self.backups:
            if backup_timestamp <= timestamp_to_restore:
                backup_to_use = (backup_timestamp, backup_data)
            else:
                break
        
        if backup_to_use is None:
            return ""
        
        backup_timestamp, backup_data = backup_to_use
        
        # Restore the database state
        self.records = {}
        for key, fields in backup_data.items():
            self.records[key] = {}
            for field, (value, remaining_ttl) in fields.items():
                if remaining_ttl is None:
                    # No expiry
                    self.records[key][field] = (value, None)
                else:
                    # Recalculate expiry time based on current timestamp
                    new_expiry = timestamp + remaining_ttl
                    self.records[key][field] = (value, new_expiry)
        
        return ""

def main():
    db = InMemoryDatabase()
    
    # Test cases from level 4 example
    queries = [
        ["SET_AT_WITH_TTL", "A", "B", "C", "1", "10"],
        ["BACKUP", "3"],
        ["SET_AT", "A", "D", "E", "4"],
        ["BACKUP", "5"],
        ["DELETE_AT", "A", "B", "8"],
        ["BACKUP", "9"],
        ["RESTORE", "10", "7"],
        ["BACKUP", "11"],
        ["SCAN_AT", "A", "15"],
        ["SCAN_AT", "A", "16"]
    ]
    
    results = []
    for query in queries:
        operation = query[0]
        if operation == "SET":
            result = db.set(query[1], query[2], query[3])
        elif operation == "GET":
            result = db.get(query[1], query[2])
        elif operation == "DELETE":
            result = db.delete(query[1], query[2])
        elif operation == "SCAN":
            result = db.scan(query[1])
        elif operation == "SCAN_BY_PREFIX":
            result = db.scan_by_prefix(query[1], query[2])
        elif operation == "SET_AT":
            result = db.set_at(query[1], query[2], query[3], int(query[4]))
        elif operation == "SET_AT_WITH_TTL":
            result = db.set_at_with_ttl(query[1], query[2], query[3], int(query[4]), int(query[5]))
        elif operation == "GET_AT":
            result = db.get_at(query[1], query[2], int(query[3]))
        elif operation == "DELETE_AT":
            result = db.delete_at(query[1], query[2], int(query[3]))
        elif operation == "SCAN_AT":
            result = db.scan_at(query[1], int(query[2]))
        elif operation == "SCAN_BY_PREFIX_AT":
            result = db.scan_by_prefix_at(query[1], query[2], int(query[3]))
        elif operation == "BACKUP":
            result = db.backup(int(query[1]))
        elif operation == "RESTORE":
            result = db.restore(int(query[1]), int(query[2]))
        
        results.append(result)
        print(f"{query}: {result}")
    
    print(f"Final results: {results}")
    # Expected: ["", "1", "", "1", "true", "1", "", "1", "B(C), D(E)", "D(E)"]

if __name__ == "__main__":
    main()